src/ai/backend/manager/sweeper/kernel.py

```diff
@@ -0,0 +1,103 @@
+from ai.backend.common.validators import TimeDelta
+from ai.backend.logging import BraceStyleAdapter
+
+from ..api.context import RootContext
+from ..config import session_hang_tolerance_iv
+from ..models import DEAD_KERNEL_STATUSES, DEAD_SESSION_STATUSES, KernelRow, SessionRow
+from .base import AbstractSweeper
+
+log = BraceStyleAdapter(logging.getLogger(__spec__.name))
+
+
+class KernelSweeper(AbstractSweeper):
+    _root_ctx: RootContext
+
+    def __init__(self, root_ctx: RootContext) -> None:
+        self._root_ctx = root_ctx
+
+    @override
+    async def sweep(self, *args) -> None:
+        query = (
+            sa.select(KernelRow)
+            .join(SessionRow, KernelRow.session_id == SessionRow.id)
+            .where(KernelRow.status.not_in(DEAD_KERNEL_STATUSES))
+            .where(SessionRow.status.in_(DEAD_SESSION_STATUSES))
+            .options(
+                noload("*"),
+                load_only(
+                    KernelRow.id,
+                    KernelRow.session_id,
+                    KernelRow.agent,
+                    KernelRow.agent_addr,
+                    KernelRow.container_id,
+                ),
+            )
+        )
+
+        async with self._root_ctx.db.begin_readonly() as conn:
+            result = await conn.execute(query)
+            kernels = result.fetchall()
+
+        kernels_per_session = defaultdict(list)
+        for kernel in kernels:
+            kernels_per_session[kernel.session_id].append(kernel)
+
+        await asyncio.gather(
+            *[
+                asyncio.create_task(
+                    self._root_ctx.registry.destroy_session_lowlevel(
+                        session_id,
+                        [
+                            {
+                                "id": kernel.id,
+                                "session_id": kernel.session_id,
+                                "agent": kernel.agent,
+                                "agent_addr": kernel.agent_addr,
+                                "container_id": kernel.container_id,
+                            }
+                            for kernel in kernels_
+                        ],
+                        reason=KernelLifecycleEventReason.HANG_TIMEOUT,
+                    )
+                )
+                for session_id, kernels_ in kernels_per_session.items()
+            ],
+            return_exceptions=False,
+        )
+
+
+@actxmgr
+async def stale_kernel_collection_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
+    session_hang_tolerance = session_hang_tolerance_iv.check(
+        await root_ctx.shared_config.etcd.get_prefix_dict("config/session/hang-tolerance")
+    )
+    default_interval_sec = 60.0
+    interval_sec = float("inf")
+    threshold: TimeDelta
+    for threshold in session_hang_tolerance["threshold"].values():
+        interval_sec = min(interval_sec, threshold.seconds)
+    if interval_sec == float("inf"):
+        interval_sec = default_interval_sec
```

Comment:
I think you should get interval as config rather than using threshold.seconds as interval.

Reply:
I've updated to fix the interval.